File: examples/relay-cascade/cascade-node-java/src/main/java/org/organicprogramming/relay/cascade/javaholon/api/PublicApi.java
package org.organicprogramming.relay.cascade.javaholon.api;

import org.organicprogramming.holons.Observability;
import relay.v1.Relay;

import java.util.Map;

public final class PublicApi {
    private PublicApi() {
    }

    public static Relay.TickResponse tick(Relay.TickRequest request) {
        Observability obs = Observability.current();
        String slug = responderSlug(obs);
        String uid = obs.cfg.instanceUid == null ? "" : obs.cfg.instanceUid;
        obs.logger("tick").info("tick received", Map.of(
                "sender", request.getSender(),
                "note", request.getNote(),
                "responder_slug", slug,
                "responder_uid", uid));
        Observability.Counter counter = obs.counter(
                "cascade_ticks_total",
                "Ticks received by this cascade node.",
                Map.of("responder_uid", uid));
        if (counter != null) {
            counter.inc();
        }
        return Relay.TickResponse.newBuilder()
                .setResponderSlug(slug)
                .setResponderInstanceUid(uid)
                .build();
    }

    static String responderSlug(Observability obs) {
        String configured = obs.cfg.slug == null ? "" : obs.cfg.slug.trim();
        return configured.isEmpty() ? "cascade-node-java" : configured;
    }
}

File: examples/relay-cascade/cascade-node-kotlin/src/main/kotlin/org/organicprogramming/relay/cascade/kotlinholon/api/PublicApi.kt
package org.organicprogramming.relay.cascade.kotlinholon.api

import org.organicprogramming.holons.Observability
import relay.v1.Relay

object PublicApi {
    fun tick(request: Relay.TickRequest): Relay.TickResponse {
        val obs = Observability.current()
        val slug = responderSlug(obs)
        val uid = obs.cfg.instanceUid
        obs.logger("tick").info(
            "tick received",
            mapOf(
                "sender" to request.sender,
                "note" to request.note,
                "responder_slug" to slug,
                "responder_uid" to uid,
            ),
        )
        obs.counter(
            "cascade_ticks_total",
            "Ticks received by this cascade node.",
            mapOf("responder_uid" to uid),
        )?.inc()
        return Relay.TickResponse.newBuilder()
            .setResponderSlug(slug)
            .setResponderInstanceUid(uid)
            .build()
    }

    private fun responderSlug(obs: Observability.Instance): String =
        obs.cfg.slug.trim().ifEmpty { "cascade-node-kotlin" }
}
